### PR TITLE
MRG: Sleep staging example

### PR DESCRIPTION
**Pull request by hubertjb**
I have started making a sleep staging example based on [this paper](https://ieeexplore.ieee.org/abstract/document/8307462/?casa_token=FVbhoYtuaIEAAAAA:PqILZbwMgZB7q_jVUHEWmUkdFS-AYnnpUM5-tysyb7qwr8lodey0kZH7GLTnezTo1K4hPsLbRFUQwg) and the [Sleep Physionet dataset](https://physionet.org/content/sleep-edfx/1.0.0/).

It's still work in progress, but feedback would already be useful!

Next steps:
* Add z-scoring of windows
* Use `crop_wake_mins` to trim Wake regions in Sleep Physionet
* Adapt hyperparameters to this low-data setting
* Plot learning curve/confusion matrix
* Add a short discussion at the end of the example on other models, refs, etc.

### CODE REVIEW

**Comment by hubertjb** [comment]
Also, this PR currently contains some changes from https://github.com/braindecode/braindecode/pull/152, so it should not be merged before the other is merged.

File: braindecode/datasets/sleep_physionet.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+import os
```
Comment (agramfort) [inline braindecode/datasets/sleep_physionet.py:1]: author and licence missing

File: braindecode/datasets/sleep_physionet.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        recordings). Can be [1], [2] or [1, 2].
+    preload: bool
+        if True, preload the data of the Raw objects.
+    load_eeg_only: bool
+        If True, only load the EEG channels and discard the others (EOG, EMG,
+        temperature, respiration) to avoid resampling the other signals.
+    crop_wake_mins: float
+        Number of minutes of wake time to keep before the first sleep event
+        and after the last sleep event. Used to reduce the imbalance in this
+        dataset. Ignore if
```
Comment (agramfort) [inline braindecode/datasets/sleep_physionet.py:34]: if?

File: braindecode/datautil/windowers.py [hunk 3]
```diff
@@ -190,15 +208,20 @@ def create_fixed_length_windows(
...
+            'to indicate end of trial/recording. Using `None`.')
+        stop_offset_samples = None
 
     list_of_windows_ds = []
     for ds in concat_ds.datasets:
-        stop = ds.raw.n_times if stop_offset_samples == 0 else stop_offset_samples
-        stop = stop - window_size_samples
+        stop = ds.raw.n_times \
+            if stop_offset_samples is None else stop_offset_samples
+        stop = stop - window_size_samples + ds.raw.first_samp
```
Comment (agramfort) [inline braindecode/datautil/windowers.py:221]:
```suggestion
        stop -= window_size_samples + ds.raw.first_samp
```
```suggestion
        stop = stop - window_size_samples + ds.raw.first_samp
```

File: braindecode/models/sleep_stager.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
+import torch
+from torch import nn
+
+
+class ChambonSleepStager(nn.Module):
```
Comment (agramfort) [inline braindecode/models/sleep_stager.py:5]: not sure about the naming here :)

File: braindecode/models/sleep_stager.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
+import torch
```
Comment (agramfort) [inline braindecode/models/sleep_stager.py:1]: author + license

File: test/unit_tests/datautil/test_windowers.py [hunk 6]
```diff
@@ -14,13 +16,15 @@
 from braindecode.datasets.moabb import fetch_data_with_moabb
 from braindecode.datautil import (
     create_windows_from_events, create_fixed_length_windows)
+from braindecode.datautil.preprocess import MNEPreproc, preprocess
```
Comment (agramfort) [inline test/unit_tests/datautil/test_windowers.py:19]:
```suggestion
from braindecode.preprocessing import MNEPreproc, preprocess
```

would be so much nicer

File: braindecode/models/sleep_stager.py (hunk 4 above)
Comment (agramfort) [reply braindecode/models/sleep_stager.py:5]: to be honest I would put the model in the example to limit the library size / scope in the future.
we can do a caffe zoo like repo later with many EEG models including pretrained ones.

what do others think?

File: braindecode/datasets/sleep_physionet.py (hunk 2 above)
Comment (agramfort) [reply braindecode/datasets/sleep_physionet.py:34]: I think you should default to the common thing ie 5 mins before and 5 mins after sleeping.

File: braindecode/models/sleep_stager.py (hunk 4 above)
Comment (hubertjb) [reply braindecode/models/sleep_stager.py:5]: I guess the current `models` directory is pretty much a zoo-like collection of EEG models (7 models so far). I'm not against moving them to a separate repo, as long as it doesn't create too much overhead. What do @robintibor @gemeinl think?

**Comment by hubertjb** [comment]
This new version is pretty complete I think.

File: braindecode/models/sleep_stager.py (hunk 4 above)
Comment (agramfort) [reply braindecode/models/sleep_stager.py]: I am not suggesting to put it another repo but in the example file. To be reconsidered when used
across files.

with all the machinery of braindecode now having 20 to 40 lines in an example to write up the module
seems an ok solution. As soon as it's in the library code changing it is harder (deprecations etc..)

Comment (robintibor) [reply braindecode/models/sleep_stager.py]: I agree with @hubertjb that models is supposed to be already like zoo-like collection of EEG models. So it depends on if the model is expected to change a lot in the near future, then ok maybe keep it in example. but if it is fairly settled I think it is nice to have it in models dir

Comment (robintibor) [reply braindecode/models/sleep_stager.py]: Similar to how torchvision has quite some models as well: https://pytorch.org/docs/stable/torchvision/models.html

**Comment by agramfort** [comment]
then we need to find a good name for the model ;)

>

**Comment by hubertjb** [comment]
> then we need to find a good name for the model ;)
> […](#)

In most papers where this architecture is used as a baseline, it's referred to as "Chambon et al. 2018". Here are some other options:
1. SleepStager (current name)
2. StagerNet (how we've called it in https://arxiv.org/pdf/2007.16104.pdf)
3. ChambonSleepStager (used in a previous version of the PR)
4. SleepStagerIEEETransNeurSys (mmh - or some variation with the journal title)
5. SleepStageClassifier

I think 1 is good, however maybe not specific enough if we end up including other sleep staging architectures in braindecode. I'd say 3 makes sense too seeing how the model is referred to in the literature.

**Comment by agramfort** [comment]
I would go with :

SleepStageClassifierChambon2018

or eventually:

SleepStagerChambon2018

**Comment by hubertjb** [comment]
I change my vote to SleepStagerChambon2018!

**Comment by robintibor** [comment]
Fine for me :)

**Comment by agramfort** [comment]
deal

>

**Comment by hubertjb** [comment]
@robintibor I think this PR is ready to be merged too!

**Review by agramfort — APPROVED** [review]
Looks good !

https://543-232335424-gh.circle-artifacts.com/0/dev/auto_examples/plot_sleep_staging.html#sphx-glr-auto-examples-plot-sleep-staging-py

**Comment by robintibor** [comment]
Thanks for the link @agramfort I saw two things:
1) 

> ### Split dataset into train and valid
> We can easily split the dataset using additional info stored in the description attribute, in this case session column. We select session_T for training and session_E for validation.

This seems from other example, please correct

2) I think for plotting it would be nicer to plot 1- balanced accuracy= balanced misclassification rate, as then curve shape is more easily comparable between misclassification rate and loss.

**Comment by hubertjb** [comment]
Here's the new version: https://545-232335424-gh.circle-artifacts.com/0/dev/auto_examples/plot_sleep_staging.html#sphx-glr-auto-examples-plot-sleep-staging-py

File: examples/plot_sleep_staging.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+
+######################################################################
+# Split dataset into train and valid
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+
+######################################################################
+# We can easily split the dataset using additional info stored in the
+# `description` attribute of :class:`braindecode.datasets.BaseDataset`,
+# in this case ``session`` column. Here, we split the examples per subject.
```
Comment (robintibor) [inline examples/plot_sleep_staging.py:147]:
```suggestion
# in this case using the ``subject`` column. Here, we split the examples per subject.
```

there is no session used here...

Comment (hubertjb) [reply examples/plot_sleep_staging.py:147]: arrh thanks for spotting that @robintibor

**Comment by robintibor** [comment]
Super! Merged!
